File: solutions/binary-search/162-find-peak-element.py
"""
# Difficulty: Medium

# 162. Find Peak Element

A peak element is an element that is strictly greater than its neighbors.

Given a 0-indexed integer array nums, find a peak element, and return its index. If the array contains multiple peaks, return the index to any of the peaks.

You may imagine that nums[-1] = nums[n] = -∞. In other words, an element is always considered to be strictly greater than a neighbor that is outside the array.

You must write an algorithm that runs in O(log n) time.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[1,2,3,1]</dd>
<dt>Output:</dt>
<dd>2 (index of peak element)</dd>
<dt>Explanation:</dt>
<dd>Peak element 4 is at index 2 in array [1,2,4,3]</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>


### METADATA:
**Techniques**: TBD
**Data Structures**: TBD
**Patterns**: TBD
**Time Complexity**: O(n)
**Space Complexity**: O(1)

### INTUITION:
A peak must exist because edges are considered smaller. Use binary search: if nums[mid] < nums[mid+1], a peak exists to the right (upward slope); otherwise a peak exists to the left or at mid (downward slope). Always converges to a peak.

### APPROACH:
1. **Initialize binary search**: Set left = 0, right = len(nums) - 1
2. **Loop until convergence**: While left < right, calculate mid = (left + right) // 2
3. **Compare with neighbor**: Check if nums[mid] > nums[mid + 1]
4. **Peak in left half**: If nums[mid] > nums[mid + 1], peak is in left half including mid, set right = mid
5. **Peak in right half**: Otherwise, peak is in right half, set left = mid + 1
6. **Converge to peak**: Continue until left == right
7. **Return peak index**: Return left (or right, they're equal) as the peak element index

### WHY THIS WORKS:
- Binary search: if nums[mid] < nums[mid+1], peak must be on right
- If nums[mid] > nums[mid+1], peak on left (mid could be peak)
- Works because adjacent elements are unequal (guaranteed peak exists)
- Doesn't need to check all elements, just finds any peak
- O(log n) time, O(1) space

### EXAMPLE WALKTHROUGH:
```
Input: nums = [1,2,3,1]
Step 1: Initialize binary search
  left = 0, right = 3

Step 2: Binary search for peak
  mid = 1: nums[1]=2 < nums[2]=3
  → Peak is on right, left = 2

  mid = 2: nums[2]=3 > nums[3]=1
  → Peak could be at mid or left, right = 2

  left = right = 2

Step 3: Check result
  nums[2] = 3 is greater than neighbors (2 and 1)

Output: 2 (index of peak element)
```

### TIME COMPLEXITY:
O(n)

### SPACE COMPLEXITY:
O(1)

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>
"""

from typing import List, Optional, Dict, Tuple


class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        """
        Finds a peak element in the array using binary search.

        Args:
            nums: List of integers where we need to find a peak element

        Returns:
            Index of any peak element in the array

        Example:
            >>> s = Solution()
            >>> s.findPeakElement([1,2,3,1])
            2
            >>> s.findPeakElement([1,2,1,3,5,6,4])
            5
        """
        # Handle edge cases
        if not nums:
            return -1
        if len(nums) == 1:
            return 0

        left, right = 0, len(nums) - 1

        # Binary search implementation
        while left < right:
            mid = left + (right - left) // 2

            # If mid element is less than next element,
            # peak must be on the right side
            if nums[mid] < nums[mid + 1]:
                left = mid + 1
            # If mid element is greater than next element,
            # peak must be on the left side or at mid
            else:
                right = mid

        return left


def test_solution() -> None:
    """
    Test cases for the solution.
    """
    solution = Solution()

    # Test case 1: Example from problem
    result = solution.findPeakElement([1, 2, 3, 1])
    expected = 2
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Single element
    result = solution.findPeakElement([1])
    expected = 0
    assert result == expected, f"Expected {expected}, got {result}"

    # Test case 3: Multiple peaks - any is valid
    result = solution.findPeakElement([1, 2, 1, 3, 5, 6, 4])
    # Result could be index 1 or 5 (both are peaks)
    assert result in [1, 5], f"Expected 1 or 5, got {result}"

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 162. Find Peak Element")

File: solutions/binary-search/162-find-peak-element.js
I'll help you create a JavaScript implementation for the Find Peak problem. I'll write a complete solution that meets the requirements and includes proper documentation.

```javascript
/**
 * Find Peak Element - JavaScript Implementation
 * LeetCode 162: https://leetcode.com/problems/find-peak-element/
 * 
 * Problem: Given an array of integers nums, find a peak element in the array.
 * A peak element is an element that is strictly greater than its neighbors.
 * 
 * Note:
 * - Array elements are 0-indexed
 * - Edge elements only need to be greater than their single neighbor
 * - It's guaranteed that the answer exists (at least one peak)
 * - Multiple peaks may exist; returning any valid peak index is acceptable
 */

/**
 * Finds a peak element in the array using binary search
 * @param {number[]} nums - Array of integers
 * @return {number} - Index of a peak element
 */
function findPeakElement(nums) {
    // Handle edge cases
    if (!nums || nums.length === 0) return -1;
    if (nums.length === 1) return 0;
    
    let left = 0;
    let right = nums.length - 1;
    
    // Binary search implementation
    while (left < right) {
        const mid = Math.floor(left + (right - left) / 2);
        
        // Compare with next element to determine which direction to search
        if (nums[mid] > nums[mid + 1]) {
            // If current element is greater than next, peak must be in left half
            right = mid;
        } else {
            // If current element is less than next, peak must be in right half
            left = mid + 1;
        }
    }
    
    // At this point, left == right and points to a peak
    return left;
}

/**
 * Helper function to test the implementation
 * @param {number[]} nums - Test input array
 */
function testFindPeak(nums) {
    const peakIndex = findPeakElement(nums);
    console.log(`Input array: [${nums.join(', ')}]`);
    console.log(`Peak found at index: ${peakIndex}`);
    console.log(`Peak value: ${nums[peakIndex]}`);
    console.log('---');
}

// Test cases
function runTests() {
    console.log('Running test cases for Find Peak Element:');
    
    // Test case 1: Simple array with one peak
    testFindPeak([1, 2, 3, 1]);
    
    // Test case 2: Array with multiple peaks
    testFindPeak([1, 2, 1, 3, 5, 6, 4]);
    
    // Test case 3: Array with peak at end
    testFindPeak([1, 2, 3, 4]);
    
    // Test case 4: Array with peak at start
    testFindPeak([4, 3, 2, 1]);
    
    // Test case 5: Single element array
    testFindPeak([1]);
}

// Execute tests if running directly
if (require.main === module) {
    runTests();
}

// Export for use in other modules
module.exports = {
    findPeakElement
};
```

This implementation includes:

1. A complete solution using binary search algorithm
2. Detailed documentation and comments explaining the approach
3. Proper error handling for edge cases
4. Test cases covering various scenarios
5. Module exports for reusability
6. Clean code structure following JavaScript conventions

The solution uses binary search to achieve O(log n) time complexity, which is more efficient than a linear scan. The algorithm works by comparing adjacent elements and moving towards the direction that's more likely to contain a peak.

Key features:
- Handles edge cases (empty array, single element)
- Works with multiple peaks (returns any valid peak)
- Efficient binary search implementation
- Comprehensive test cases
- Clear documentation and comments

The code can be run directly to execute the test cases or imported as a module in other files. The test cases demonstrate the functionality with various input scenarios.